openflow13/nx_action.go

```diff
@@ -0,0 +1,956 @@
+	n += len(b)
+	binary.BigEndian.PutUint16(data[n:], a.InPort)
+	n += 2
+	data[n] = a.TableID
+	n += 1
+	copy(data[n:], a.pad)
+	n += 3
+
+	return
+}
+
+func (a *NXActionResubmitTable) UnmarshalBinary(data []byte) error {
+	n := 0
+	a.NXActionHeader = new(NXActionHeader)
+	err := a.NXActionHeader.UnmarshalBinary(data[n:])
+	n += int(a.NXActionHeader.Len())
+	if len(data) < int(a.Len()) {
+		return errors.New("the byte array has wrong size to unmarshal an NXActionResubmitTable message")
+	}
+	a.InPort = binary.BigEndian.Uint16(data[n:])
+	n += 2
+	a.TableID = data[n]
+	n += 1
+	copy(a.pad, data[n:n+3])
+
+	return err
+}
+
+func NewNXActionResubmitTableCT(inPort uint16, tableId uint8) *NXActionResubmitTable {
+	a := newNXActionResubmitTableCT()
+	a.InPort = inPort
+	a.TableID = tableId
+	return a
+}
+
+func NewNXActionResubmitTableCTNoInPort(tableId uint8) *NXActionResubmitTable {
+	a := newNXActionResubmitTableCT()
+	a.InPort = OFPP_IN_PORT
+	a.TableID = tableId
+	return a
+}
+
+// nxast_resubmit_table
+type NXActionCTNAT struct {
+	*NXActionHeader
+	pad          []byte // 2 bytes
+	Flags        uint16 // nat Flags to identify snat/dnat, and nat algorithms: random/protocolHash, connection persistent
+	rangePresent uint16 // mark if has set nat range, including ipv4/ipv6 range, port range
+
+	rangeIPv4Min  net.IP
+	rangeIPv4Max  net.IP
+	rangeIPv6Min  net.IP
+	rangeIPv6Max  net.IP
+	rangeProtoMin *uint16
+	rangeProtoMax *uint16
+}
+
+func NewNXActionCTNAT() *NXActionCTNAT {
+	a := new(NXActionCTNAT)
+	a.NXActionHeader = NewNxActionHeader(NXAST_NAT)
+	a.Length = 16
+	a.pad = make([]byte, 2, 2)
+	return a
+}
+
+func (a *NXActionCTNAT) Len() (n uint16) {
+	a.Length = uint16(math.Ceil(float64(a.Length)/float64(8))) * 8
+	return a.Length
+}
+
+func (a *NXActionCTNAT) MarshalBinary() (data []byte, err error) {
+	optData := make([]byte, 0)
+	optN := 0
+	if a.rangeIPv4Min != nil {
+		optData = append(optData[optN:], a.rangeIPv4Min.To4()...)
+		optN += 4
+	}
+	if a.rangeIPv4Max != nil {
+		optData = append(optData[optN:], a.rangeIPv4Max.To4()...)
+		optN += 4
+	}
+	if a.rangeIPv6Min != nil {
+		optData = append(optData[optN:], a.rangeIPv6Min.To16()...)
+		optN += 16
+	}
+	if a.rangeIPv6Max != nil {
+		optData = append(optData[optN:], a.rangeIPv6Max.To16()...)
+		optN += 16
+	}
+	if a.rangeProtoMin != nil {
+		binary.BigEndian.PutUint16(optData[optN:], *a.rangeProtoMin)
+		optN += 2
+	}
+	if a.rangeProtoMin != nil {
+		binary.BigEndian.PutUint16(optData[optN:], *a.rangeProtoMax)
+		optN += 2
+	}
+
+	data = make([]byte, a.Len())
+	b := make([]byte, a.NXActionHeader.Len())
+	n := 0
+
+	b, err = a.NXActionHeader.MarshalBinary()
+	copy(data[n:], b)
+	n += len(b)
+	copy(data[n:], a.pad)
+	n += 2
+	binary.BigEndian.PutUint16(data[n:], a.Flags)
+	n += 2
+	binary.BigEndian.PutUint16(data[n:], a.rangePresent)
+	n += 2
+	copy(data[n:], optData)
+	n += optN
+
+	// padding the message if the length is not an integer multiple of 8 bytes
+	if n < int(a.Length) {
```

Comment:
I am not sure I understand the point of this TBH: if `data` already has the correct length , there is nothing you need to do. You already have the necessary padding.

Reply:
If padding is the last part, it might be right, but for some messages, padding is in the middle, I think padding should be needed. Or do you think it will work if I could only move index of byte array if padding is in the middle? Anyway I could try to save this operation of "copy padding".

Reply:
The new version looks better to me. Thanks.